File: ircd/tests/stress.py
import tornado.ioloop
import tornado.iostream
import socket
import random
import sys

address = ('localhost', 5556)


class Client(object):
    def __init__(self):
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
        self.stream = tornado.iostream.IOStream(s)
        self.stream.connect(address, self.on_connect)
        self.stream.read_until('\n', self.on_data)
        self.chans = set()

        self.timer = tornado.ioloop.PeriodicCallback(self.do_something, 500)
        self.timer.start()

    def send(self, *args):
        self.stream.write(' '.join(args) + '\r\n')

    def on_connect(self):
        self.send('NICK py' + str(random.randint(0, 1000)))
        self.send('USER u\n')

    def on_data(self, data):
        self.stream.read_until('\n', self.on_data)

    def do_something(self):
        r = random.random()
        if len(self.chans) < 3:
            r = 0

        if r < 0.1:
            ch = '#' + str(random.randint(0, 10))
            self.send('JOIN', ch)
            self.chans.add(ch)
        elif r < 0.2:
            ch = random.choice(list(self.chans))
            self.send('PART', ch)
            self.chans.remove(ch)
        else:
            ch = random.choice(list(self.chans))
            self.send('PRIVMSG', ch, ':heloooooo nurse')


for _ in xrange(int(sys.argv[1])):
    Client()

tornado.ioloop.IOLoop.instance().start()

File: ircd/tests/stress.js
var WebSocket = require('ws');

var choice = function (a) {
    var r = (Math.random()*2000000000)|0;
    return a[r % a.length];
};

var new_client = function() {
    var ws = new WebSocket('ws://localhost:8888/sjs/websocket');
    var chans = [];

    var send = function() {
        var args = Array.prototype.slice.call(arguments);
        var data = args.join(' ') + '\r\n';
        ws.send(data);
        //console.log('> ' + data.trim());
    };

    ws.on('open', function() {
        send('NICK zxc' + ((Math.random()*1000)|0));
        send('USER u');
    });

    ws.on('message', function(data, flags) {
        //console.log('< ' + data.trim());
    });

    var do_something = function() {
        r = Math.random();

        if (chans.length < 3)
            r = 0;

        if (r < 0.1) {
            ch = '#' + ((Math.random()*10)|0);
            send('JOIN', ch);
            chans.push(ch);
        }
        else if (r < 0.2) {
            ch = choice(chans);
            send('PART', ch);
            chans.splice(chans.indexOf(ch), 1);
        }
        else {
            ch = choice(chans);
            send('PRIVMSG', ch, ':heloooooo nurse');
        }
    };

    setInterval(do_something, 500);
};

var n = process.argv[2]|0;
while (n--)
    new_client();
